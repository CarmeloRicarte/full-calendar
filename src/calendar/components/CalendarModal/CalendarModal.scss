$color_1: rgb(51, 51, 51);
$background-color_1: rgba(0, 0, 0, 0.3);

.ReactModalPortal {
  > div {
    opacity: 0;
  }
  .ReactModal__Overlay {
    align-items: center;
    display: flex;
    justify-content: center;
    transition: opacity 0.2s ease-in-out;
    z-index: 999;
  }
  .ReactModal__Overlay--after-open {
    opacity: 1;
  }
  .ReactModal__Overlay--before-close {
    opacity: 0;
  }
}
.modal-fondo {
  background-color: $background-color_1;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  position: fixed;
}
.modal {
  background: white;
  border-radius: 5px;
  color: $color_1;
  display: inline;
  max-height: 620px;
  max-width: 500px;
  outline: none;
  padding: 10px;
}
