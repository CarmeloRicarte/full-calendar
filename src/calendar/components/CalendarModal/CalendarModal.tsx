import { useState } from "react";
import Modal from "react-modal";
import { addHours } from "date-fns";
import DatePicker, { registerLocale } from "react-datepicker";
import es from "date-fns/locale/es";

import "./CalendarModal.scss";
import "react-datepicker/dist/react-datepicker.css";
import { CalendarEvent } from "../../interfaces";

registerLocale("es", es);

const customStyles = {
  content: {
    top: "50%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)",
  },
};

Modal.setAppElement("#root");

const formData: CalendarEvent = {
  title: "",
  notes: "",
  start: new Date(),
  end: addHours(new Date(), 2),
};

export const CalendarModal = () => {
  const [isOpen, setIsOpen] = useState(true);
  const [formValues, setFormValues] = useState<CalendarEvent>(formData);

  const onCloseModal = () => {
    setIsOpen(false);
  };

  const onInputChange = ({
    target,
  }: {
    target: HTMLInputElement | HTMLTextAreaElement;
  }) => {
    setFormValues({
      ...formValues,
      [target.name]: target.value,
    });
  };

  const onDateChange = (date: Date, changing: "start" | "end") => {
    setFormValues({
      ...formValues,
      [changing]: date,
    });
  };

  return (
    <Modal
      isOpen={isOpen}
      onRequestClose={onCloseModal}
      style={customStyles}
      contentLabel="Example Modal"
      className="modal"
      overlayClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1> Nuevo evento </h1>
      <hr />
      <form className="container">
        <div className="form-group mb-2">
          <label>Fecha y hora inicio</label>
          <DatePicker
            locale="es"
            timeCaption="Hora"
            selected={formValues.start}
            className="form-control"
            onChange={(date: Date) => onDateChange(date, "start")}
            dateFormat="Pp"
            showTimeSelect
          />
        </div>

        <div className="form-group mb-2">
          <label>Fecha y hora fin</label>
          <DatePicker
            locale="es"
            timeCaption="Hora"
            minDate={formValues.start}
            selected={formValues.end}
            className="form-control"
            onChange={(date: Date) => onDateChange(date, "end")}
            dateFormat="Pp"
            showTimeSelect
          />
        </div>

        <hr />
        <div className="form-group mb-2">
          <label>Titulo y notas</label>
          <input
            type="text"
            className="form-control"
            placeholder="Título del evento"
            name="title"
            autoComplete="off"
            value={formValues.title}
            onChange={onInputChange}
          />
          <small id="emailHelp" className="form-text text-muted">
            Una descripción corta
          </small>
        </div>

        <div className="form-group mb-2">
          <textarea
            className="form-control"
            placeholder="Notas"
            rows={5}
            name="notes"
            value={formValues.notes}
            onChange={onInputChange}
          ></textarea>
          <small id="emailHelp" className="form-text text-muted">
            Información adicional
          </small>
        </div>

        <button type="submit" className="btn btn-outline-primary btn-block">
          <i className="far fa-save"></i>
          <span> Guardar</span>
        </button>
      </form>
    </Modal>
  );
};
