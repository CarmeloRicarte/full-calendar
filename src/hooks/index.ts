export * from "./dispatchSelector";
export * from "./useCalendarStore";
export * from "./useForm";
export * from "./useUiStore";
