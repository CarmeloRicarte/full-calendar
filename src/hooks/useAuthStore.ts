import { useSelector } from "react-redux";
import { calendarApi } from "../api";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
  RootState,
} from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector(
    (state: RootState) => state.auth
  );
  const dispatch = useAppDispatch();

  const startLogin = async ({
    email,
    password,
  }: {
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth", {
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error) {
      dispatch(onLogout("Credenciales incorrectas"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const startRegister = async ({
    name,
    email,
    password,
  }: {
    name: string;
    email: string;
    password: string;
  }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/new", {
        name,
        email,
        password,
      });
      setTokenLocalStorage(data.token);

      dispatch(
        onLogin({
          name: data.name,
          uid: data.uid,
        })
      );
    } catch (error: any) {
      if (error.response.data?.msg) {
        dispatch(onLogout(error.response.data?.msg));
      } else {
        const keyFirstElement = Object.keys(error.response.data?.errors)[0];
        keyFirstElement &&
          dispatch(onLogout(error.response.data?.errors[keyFirstElement]?.msg));
      }

      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 1000);
    }
  };

  const setTokenLocalStorage = (token: string) => {
    localStorage.setItem("token", token);
    localStorage.setItem(
      "token-expires-date",
      new Date().getTime() + 2 * 60 * 60 * 1000 + ""
    );
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogout());
    try {
      const { data } = await calendarApi.get("/auth/renew");

      // logout user if token expired
      const expirationTokenDate = localStorage.getItem("token-expires-date");
      if (expirationTokenDate && +expirationTokenDate < new Date().getTime()) {
        dispatch(onLogout());
      } else {
        setTokenLocalStorage(data);
        dispatch(
          onLogin({
            name: data.name,
            uid: data.uid,
          })
        );
      }
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  return {
    //* Properties
    status,
    user,
    errorMessage,
    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
  };
};
