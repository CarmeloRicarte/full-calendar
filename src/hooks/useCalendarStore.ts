import { useAppDispatch } from "./dispatchSelector";
import { useSelector } from "react-redux";
import { onSetActiveEvent, RootState } from "../store";
import { CalendarEvent } from "../calendar/interfaces";

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useSelector(
    (state: RootState) => state.calendar
  );

  const setActiveEvent = (calendarEvent: CalendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  return {
    //* Properties
    events,
    activeEvent,
    //*Methods
    setActiveEvent,
  };
};
