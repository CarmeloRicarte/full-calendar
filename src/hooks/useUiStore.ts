import { useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal, RootState } from "../store";
import { useAppDispatch } from "./dispatchSelector";

export const useUiStore = () => {
  const dispatch = useAppDispatch();
  const { isDateModalOpen } = useSelector((state: RootState) => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => dispatch(onCloseDateModal());

  return {
    //* Properties
    isDateModalOpen,

    //*Methods
    openDateModal,
    closeDateModal,
  };
};
