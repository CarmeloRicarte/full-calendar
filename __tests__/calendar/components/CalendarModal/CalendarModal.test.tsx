import { describe, test, expect, vi, beforeEach } from "vitest";
import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import * as useCalendarStore from "../../../../src/hooks/useCalendarStore";
import * as useUiStore from "../../../../src/hooks/useUiStore";
import { CalendarModal } from "../../../../src/calendar/components/CalendarModal/CalendarModal";
import { events } from "../../../__fixtures__/calendarStates";

vi.mock("../../../../src/hooks/useCalendarStore");
vi.mock("../../../../src/hooks/useUiStore");

describe("Tests of <CalendarModal />", () => {
  const mockStartSavingEvent = vi.fn();
  const mockCloseDateModal = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test("should render the modal", () => {
    (useCalendarStore as any).useCalendarStore = vi.fn().mockReturnValue({
      activeEvent: null,
    });
    (useUiStore as any).useUiStore = vi.fn().mockReturnValue({
      isDateModalOpen: true,
    });

    const { container } = render(<CalendarModal />);
    expect(container).toMatchSnapshot();
  });

  test("should render the modal with the active event", () => {
    (useCalendarStore as any).useCalendarStore = vi.fn().mockReturnValue({
      activeEvent: events[0],
    });
    (useUiStore as any).useUiStore = vi.fn().mockReturnValue({
      isDateModalOpen: true,
    });
    render(<CalendarModal />);
    const startDateInput = screen.getByDisplayValue("19/12/2022, 13:00");
    const endDateInput = screen.getByDisplayValue("19/12/2022, 15:00");
    const titleInput = screen.getByDisplayValue("Event 1");
    const notesInput = screen.getByDisplayValue("Some note");

    expect(screen.getByRole("heading", { level: 1 }).textContent).toContain(
      "Editar evento"
    );
    expect({
      startDateInput,
      endDateInput,
      titleInput,
      notesInput,
    }).toBeTruthy();
  });
});
