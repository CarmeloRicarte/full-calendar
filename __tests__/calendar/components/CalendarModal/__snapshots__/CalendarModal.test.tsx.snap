// Vitest Snapshot v1

exports[`Tests of <CalendarModal /> > should render the modal 1`] = `<div />`;
